### DISCUSSION TITLE
Why does this have many diffrent options to add a control to another?

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [alternetsoft/AlternetUI#139] [Ideas]
- Window.Menu
- Window.Children
- Window.Content
- Window.StatusBar
- Control.Add
- Control.Add*
- Menu.Items
- MenuItem.Items

These should be unified.
Right now it's really hard to build ui in C# because if we use the wrong method it doesn't get laid out properly.

**Comment by generalloki** [comment]
I am moving this to discussions. Here is the proper way to add one control to another:

```cs
            var editor = new TextBox();
            editor.Parent = someParent;
```

Also you can disable/enable layout:

```cs
            someParent.SuspendLayout();

            var editor = new TextBox();
            editor.Parent = someParent;

            var button = new Button();
            button.Parent = someParent;
            
            someParent.ResumeLayout();

```

**Reply by Shadowblitz16** [reply]
@generalloki can you remove the other ways and force a parent parameter as the first argument to a controls constructor?

```cs

public class Control 
{
      // can't be constructed without parent.
      public Control(Control parent) {}
     
      // for application root
      internal Control() {}
}
```

that would clean up the api and ensure things are initialized.

forms would be...
```cs
public class Form : Control
{
      // not mdi just ownership
      public Form(Form parent) : base(parent)
      {

      }
      
      // calls internal Control constructor
      public Form(Application app)
      {
    
      }
}

```

**Comment by Shadowblitz16** [comment]
@generalloki that's not very scalable when building ui up in C#.
something like this would be better...
```cs
Control Build() => new StackPanel()
{   
    Orentation = Orentation.Vertical,
    Children = 
    {
        new Menu() 
        {
            Children = 
            {
                new MenuItem("_File") = 
                {
                    Children = 
                    {
                        new MenuItem("New", OnFileNew),
                        new Seperator()
                    }
                }
            }
        },
        new Panel()
        {
            UserPaint = true,
            Background = Brush.Black
        }
    }
};
```
or better yet something like this...

```cs
Control Build() => new StackPanel
(
    Orentation.Vertical,
    new Menu
    (
        new MenuItem
        (
            "_File",
            new MenuItem("New", OnFileNew),
            new Seperator(Orentation.Vertical)
        )
    )
    new Panel().WithBackground(Brush.Black).WithUserPainting()
)
```

also when you have 12 different ways of adding children it's not immediately clear to the user which way is correct for which control.

**Comment by generalloki** [comment]
@Shadowblitz16 

We will add control's and form constructors with parent parameter. I will post here an update when it's done.

As to other proposals, we will think about it. Let's progress on this issue step by step.

**Comment by generalloki** [comment]
@Shadowblitz16 
Added to all Control descendants: new constructor with parameter which allows to specify parent of the control.
But not for Window descendants and not for Menu, MenuItem, MainMenu. Menu related classes will be rewritten in the near future, they will be derived from FrameworkElement instead of Control.

**Comment by Shadowblitz16** [comment]
@generalloki I noticed doing this does not show a back window...
```cs
public class Editor : Window
{
    public Editor()
    {
        InitializeComponent();
    }

    private void InitializeComponent()
    {
        Title  = "New SMBX Editor";
        Width  = 800;
        Height = 600;
        //Menu   = BuildMainMenu();
        BuildContent().Parent = this;
    }

    private Control BuildContent()
    {
        var content = new Panel();
        content.VerticalAlignment   = VerticalAlignment.Fill;
        content.HorizontalAlignment = HorizontalAlignment.Fill;
        content.UserPaint  = true;
        content.Background = new SolidBrush(Color.Black);
        return content;
    }

```

The panel is set to fill on both axis, user paint is true and background is set to black.
I then use the `Parent` property to set it as the content for the window but it doesn't show up.

![image](https://github.com/user-attachments/assets/db4ba550-4d52-4e18-9e4f-e5d8890b2a15)

I have tried setting the alignment to stretch and user paint to false.

**Reply by generalloki** [reply]
I have checked the code that you have provided and it works fine with the master version. Here is the picture:

![image](https://github.com/user-attachments/assets/5fd1b527-e403-48ee-92e6-8b6b269bcdd0)

**Reply by Shadowblitz16** [reply]
ok I was using the latest nuget prerelease

**Comment by generalloki** [comment]
Also some info in https://github.com/alternetsoft/AlternetUI/issues/192
